modeling/modeling-mongodb-documents-with-native-go-data-structures.md

```diff
@@ -0,0 +1,164 @@
+
+In each of the tutorials, we made use of `bson.A`, `bson.D`, and `bson.M`, which represent arrays, documents, and maps. However, these are primitive data structures that are part of the MongoDB Go driver, and not necessarily the best way for interacting with data, both within the application and the database.
+
+We're going to look at an alternative way to interact with data through the MongoDB Go driver operations. This time we're going to map MongoDB document fields to native Go data structures.
+
+## The Requirements
+
+To be successful with this tutorial, the following should be met:
+
+- A MongoDB Atlas cluster
+- Go 1.10+
+
+You should already have Go installed and configured as this tutorial won't explain how to do this. You should also have your [MongoDB Atlas](https://www.mongodb.com/cloud) cluster properly configured with the correct whitelisting. If you need help doing this, check out my [previous tutorial](https://www.mongodb.com/blog/post/quick-start-golang--mongodb--starting-and-setup) on the subject.
+
+> While MongoDB Atlas has a forever FREE tier, apply promotional code [NICRABOY200](https://www.mongodb.com/cloud) to receive credit towards a more powerful cluster.
+
+If you haven't experienced the previous tutorials in the series, viewing them would be beneficial, but not a requirement.
+
+## Creating a Go Data Structure with BSON Annotations
+
+The premise of this tutorial is to use a native Go data structure when working with MongoDB, but let's take a step back and look at what we were working with in the past few tutorials.
+
+You might remember something like the following:
+
+```go
+bson.M{
+    "title": "The Polyglot Developer Podcast",
+    "author": "Nic Raboy",
+    "tags": bson.A{"development", "programming", "coding"}
+}
+```
+
+Working with a `bson.M` will leave you with a `map[string]interface{}` which isn't the most complicated in the world, but isn't necessarily the best for all scenarios in my opinion. Things get more challenging as you start working with `bson.D` as well.
+
+A lot of these scenarios can be simplified by mapping the fields of the document to fields of a data structure, similar to how Go does it with JSON and XML.
+
+Take the following Go data structure:
+
+```go
+type Podcast struct {
+	ID     primitive.ObjectID `bson:"_id,omitempty"`
+	Title  string             `bson:"title,omitempty"`
+	Author string             `bson:"author,omitempty"`
+	Tags   []string           `bson:"tags,omitempty"`
+}
+```
+
+The above data structure is nearly identical to the `bson.M` that was used previously, with the exception that it has an `ID` field. You'll notice the BSON annotations. These annotations are the actual fields that would appear in the document within MongoDB. The `omitempty` means that if there is no data in the particular field, when saved to MongoDB the field will not exist on the document rather than existing with an empty value.
```

Comment:
I would also mention the benefits of typing out the data structures. For example having autocomplete, being able to write methods specific to the struct, etc. This will greatly help highlight why structs are superior to just using the primitive bson types to represent our data. :)